### PR TITLE
fix(#6962): Fix race conditions in VWindowItem.ts during transition.

### PR DESCRIPTION
**Pull request by chouhanyang**
From time to time, the transition state of the VWindow container is
not finished and the height will be fixed. This change fixed this
problem.

FINDINGS:

- If transitionend fired before nextTick(), then done() is never
  going to be invoked.
- onEnter() itself may be re-entered during transiton. Then the
  state is messed up.
- onBeforeLeave() may be called after cancel, then the height
  is stuck.
- onBeforeEnter() may be re-entered after first call. Calling
  this multiple times also wrongly reset transition state.

SOLUTION:

- Put correct guards on each event handler. Check of isActive and
  wasCancelled accordingly for each event to prevent re-entrant.
- Put a timer in onEnter() to ensure done() is always invoked and
  deactivate() after timeout if anything goes wrong.

TESTS:

- Add log instrumentations for observation.
- Test quickly switching tabs and see container always deactivate.

## Markup:

<details>

```vue
// Paste your FULL Playground.vue here
```
</details>

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)
- [x] I've added new examples to the kitchen (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by chouhanyang** [comment]
Hmm... I think I have fixed all broken tests. However, the "radioOn" breakage was caused by previous commits. How can I solve this? Any pointer?

**Comment by chouhanyang** [comment]
Never mind, all checks passed. Please review my code. Thank you!

**Comment by chouhanyang** [comment]
Weird. Travis CI job passed. But github shows pending. I'll force update again to see if I can trigger the check to run again.

**Comment by johnleider** [comment]
This breaks the tab content height transition.

**Expected:**
expected

**Actual:**
actual

**Comment by chouhanyang** [comment]
> This breaks the tab content height transition.
> 

Hi John,
Thanks for pointing that out. Now I understand what the business of setting clientHeight is. I have investigated more and have tried several different approach to fix this issue:

1. I tried to put back onBeforeLeave(), but this is triggered on different item. And it may be fired before or after onBeforeEnter(), which depends on if we move forward the tabs or backward. This is bad because in the case of switch tabs continuously, onBeforeLeave() may fire very late, which cause the height to stuck.

2. I also tried to set initial height within onBeforeEnter() where we kick off the transition process. I use the current height of the parent element as initial height. However, it works only if we walk backward of the tabs. When walking forward the tabs, parent element collapse to 0px height. The root cause may be similar to 1. Where I tried to fix the order where we set isActive in VItemGroup.ts, but no use. Vue just fired with its own order.

3. Finally, I have to add an beforeChange() from VItemGroup.ts. So in this callback, we can make sure no transition events are fired. Then we can obtain current height correctly and start transition height inside onBeforeEnter(). 

This change is a bit more involved than I originally though. Appreciate if you can take another look and let me know if this callback is something that fits the coding standard.

**Comment by johnleider** [comment]
For the most part everything looks good. I did uncover an issue with the playground of `v-tabs` if you could take a look.

tabs

**Comment by chouhanyang** [comment]
John,

Thanks for taking another look. Just to clarify, the problem you mentioned here is that the content breaks out of its container transiently, right?

If so, this is related to another issue #7102 where I find the problem is because the active state terminated early due to numbers of reasons.

But my change could make this issue worse because the change preemptively remove the active state while cancel. I'll see if I can also fix #7102 by allowing multiple transition tabs running at the same time.

**Comment by chouhanyang** [comment]
Good news!

The reason why onAfterEnter() was skipped is simply it was cancelled. With the new change we allow 2 windows to have their own transition states individually. And then, we can use a counter activeWindows as a semiphore to capture entire transition duration with multiple cancellation in between.

- Use activeWindows to deduce if the container should be in active state.
- Compute isActive instead.
- Got rid of transitionEnd() event. 
- Let cancel go to the same path as after enter/leave path.
- Got rid of the done() callback related code and the timer in previous change.
- Got rid of all async calls requestAnimationFrame() since it's no longer necessary.

So basically as long as we handle the following 4 paths correctly. We don't need any hacks.

1. Normal enter: onBeforeEnter() .... onAfterEnter().
2. Cancel enter: onBeforeEnter() ... onEnterCancelled().
3. Normal leave: onBeforeLeave() ... onAfterLeave().
4. Cancel leave: onBeforeLeave() ... onLeaveCancelled().

Every handler is now synchronous. I'm pretty happy with the new change since it is a lot simpler than previous attempts. I'm doing more tests in the mean time. Let me know what you think.

**Comment by chouhanyang** [comment]
Since now we handle both enter and leave paths. We can detect the first onBeforeXXX() event easily. So we can get rid of the beforeChange() event and the trigger from VItemGroup.ts. It's much cleaner now.

**Comment by chouhanyang** [comment]
@johnleider 

Thanks John. I think you can close issue #7102 as well. The root causes are the same.
